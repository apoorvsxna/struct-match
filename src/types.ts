import { SyntaxNode } from 'tree-sitter';

/** An alias for a tree-sitter `SyntaxNode`. */
export interface ASTNode extends SyntaxNode {}

/** A map of captured metavariable names to their string values. */
export interface Metavariable {
  [key: string]: string;
}

/** A single match result from a search query. */
export interface Match {
  /** The id of the rule that produced this match. */
  ruleId: string;
  /** The AST node that matched. */
  node: ASTNode;
  /** The text content of the matched node. */
  matchedText: string;
  startLine: number;
  endLine: number;
  /** Captured metavariables (e.g., `{ "$NAME": "user" }`). */
  metavariables: Metavariable;
}

/**
 * A declarative rule for querying the AST.
 * Rules can be nested to specify context.
 */
export interface RulePattern {
  /** The core tree-sitter pattern string. */
  pattern?: string;
  /** Matches any of the provided patterns. */
  any?: RulePattern[];
  /** Narrows the search to descendants of a node matching this pattern. */
  inside?: RulePattern;
  /** Requires the matched node to contain a node matching this pattern. */
  contains?: RulePattern;
  /** Requires a preceding sibling to match this pattern. */
  precedes?: RulePattern;
  /** Requires a following sibling to match this pattern. */
  follows?: RulePattern;
  /** Inverts the match; excludes nodes that match this pattern. */
  not?: RulePattern;
}