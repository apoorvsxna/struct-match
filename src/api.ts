import { Match, RulePattern } from './types';
import { CodeParser } from './core/Parser';
import { Matcher } from './core/Matcher';
import { RuleProcessor } from './core/RuleProcessor';
import { getLineNumbers, getMatchText } from './utils/NodeAnalyzer';
import yaml from 'js-yaml';

const codeParser = new CodeParser();
const matcher = new Matcher(codeParser);
const ruleProcessor = new RuleProcessor(matcher);

/**
 * Finds all occurrences of a simple pattern string in the given source code.
 * @param sourceCode The source code to analyze.
 * @param pattern The pattern to search for
 * @returns A promise that resolves to an array of all found matches.
 */
export async function findByPattern(
  sourceCode: string,
  pattern: string
): Promise<Omit<Match, 'ruleId'>[]> {
  const sourceTree = codeParser.parse(sourceCode);
  const rule: RulePattern = { pattern };

  const intermediateMatches = matcher.find(sourceTree, rule);

  return intermediateMatches.map(({ node, metavariables }) => {
    const { startLine, endLine } = getLineNumbers(sourceCode, node);
    const matchedText = getMatchText(node);

    return {
      node,
      matchedText,
      startLine,
      endLine,
      metavariables,
    };
  });
}

/**
 * Finds all matches in the source code based on a YAML rule
 * or an array of rules.
 * @param sourceCode The source code to analyze.
 * @param yamlRule A string containing one or more rules in YAML format.
 * @returns A promise that resolves to an array of all found matches.
 */
export async function findByRule(
  sourceCode: string,
  yamlRule: string
): Promise<Match[]> {
  const sourceTree = codeParser.parse(sourceCode);
  const loadedYaml = yaml.load(yamlRule);

  const rules = Array.isArray(loadedYaml) ? loadedYaml : [loadedYaml];

  const allMatches: Match[] = [];

  for (const fullRule of rules) {
    // Silently skip any malformed entries in the array.
    if (!fullRule || typeof fullRule.id !== 'string' || typeof fullRule.rule !== 'object') {
      continue;
    }

    const ruleId = fullRule.id;
    const rulePatternAsYaml = yaml.dump(fullRule.rule);

    const intermediateMatches = await ruleProcessor.process(
      sourceTree,
      rulePatternAsYaml
    );

    const matchesForThisRule = intermediateMatches.map(({ node, metavariables }) => {
      const { startLine, endLine } = getLineNumbers(sourceCode, node);
      const matchedText = getMatchText(node);

      return {
        ruleId,
        node,
        matchedText,
        startLine,
        endLine,
        metavariables,
      };
    });

    allMatches.push(...matchesForThisRule);
  }

  return allMatches;
}