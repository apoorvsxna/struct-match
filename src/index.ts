/**
 * This is the main entry point for your library.
 * It exports the public-facing API functions and the core data types.
 */
export * from './api';
export * from './types';