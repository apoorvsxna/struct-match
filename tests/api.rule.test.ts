import { findMatchesByRule } from '../src/api';
import { ruleCases } from './cases/rule-cases';
import yaml from 'js-yaml';

describe('findMatchesByRule', () => {
  ruleCases.forEach(testCase => {
    it(`should correctly handle: ${testCase.name}`, async () => {
      // We just need to dump the rule object to a YAML string.
      const yamlStringToTest = yaml.dump(testCase.rule);

      const matches = await findMatchesByRule(testCase.sourceCode, yamlStringToTest);

      const simplifiedMatches = matches.map(match => ({
        matchedText: match.matchedText,
        metavariables: match.metavariables,
      }));

      expect(simplifiedMatches).toMatchSnapshot();
    });
  });
});
