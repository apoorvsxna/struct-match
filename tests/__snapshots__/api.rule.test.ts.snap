// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`findByRule should correctly handle: Composite Rule - any 1`] = `
[
  {
    "endLine": 2,
    "matchedText": "const x = 5;",
    "metavariables": {
      "$X": "x",
      "$Y": "5",
    },
    "ruleId": "composite-any",
    "startLine": 2,
  },
  {
    "endLine": 4,
    "matchedText": "console.log("random")",
    "metavariables": {
      "$X": ""random"",
    },
    "ruleId": "composite-any",
    "startLine": 4,
  },
]
`;

exports[`findByRule should correctly handle: Context Specifier - contains 1`] = `
[
  {
    "endLine": 4,
    "matchedText": "function go() {
          console.log("random");
      }",
    "metavariables": {
      "$X": "go",
    },
    "ruleId": "context-contains",
    "startLine": 2,
  },
]
`;

exports[`findByRule should correctly handle: Context Specifier - follows 1`] = `
[
  {
    "endLine": 4,
    "matchedText": "console.log(x)",
    "metavariables": {},
    "ruleId": "context-follows",
    "startLine": 4,
  },
]
`;

exports[`findByRule should correctly handle: Context Specifier - inside 1`] = `
[
  {
    "endLine": 3,
    "matchedText": "console.log("random")",
    "metavariables": {},
    "ruleId": "context-inside",
    "startLine": 3,
  },
]
`;

exports[`findByRule should correctly handle: Context Specifier - not 1`] = `
[
  {
    "endLine": 2,
    "matchedText": "console.log("goodbye")",
    "metavariables": {},
    "ruleId": "context-not",
    "startLine": 2,
  },
]
`;

exports[`findByRule should correctly handle: Context Specifier - precedes 1`] = `
[
  {
    "endLine": 2,
    "matchedText": "console.log(x)",
    "metavariables": {},
    "ruleId": "context-precedes",
    "startLine": 2,
  },
]
`;

exports[`findByRule should correctly handle: Metavariable consistency across context specifiers 1`] = `
[
  {
    "endLine": 5,
    "matchedText": "function go() {
          var text = "random";
          console.log(text);
      }",
    "metavariables": {
      "$FUNC": "go",
    },
    "ruleId": "metavariable-consistency",
    "startLine": 2,
  },
]
`;

exports[`findByRule should correctly handle: Metavariable consistency across context specifiers 2`] = `
[
  {
    "endLine": 5,
    "matchedText": "function go() {
          var text = "random";
          console.log(text);
      }",
    "metavariables": {
      "$FUNC": "go",
    },
    "ruleId": "rule1",
    "startLine": 2,
  },
  {
    "endLine": 4,
    "matchedText": "console.log(text)",
    "metavariables": {
      "$A": "text",
    },
    "ruleId": "rule2",
    "startLine": 4,
  },
]
`;

exports[`findByRule should correctly handle: Multiple context specifiers (nested) 1`] = `
[
  {
    "endLine": 5,
    "matchedText": "function go() {
          var text = "random";
          console.log(text);
      }",
    "metavariables": {
      "$FUNC": "go",
    },
    "ruleId": "multiple-specifiers-nested",
    "startLine": 2,
  },
]
`;

exports[`findByRule should correctly handle: Multiple context specifiers (non-nested) 1`] = `
[
  {
    "endLine": 4,
    "matchedText": "console.log("random")",
    "metavariables": {},
    "ruleId": "multiple-specifiers-non-nested",
    "startLine": 4,
  },
]
`;

exports[`findByRule should correctly handle: Single Pattern, Direct 1`] = `
[
  {
    "endLine": 2,
    "matchedText": "const x = 5;",
    "metavariables": {},
    "ruleId": "single-pattern",
    "startLine": 2,
  },
]
`;
